src/main/java/org/candlepin/model/OwnerCurator.java

```diff
@@ -189,55 +199,55 @@ public Owner getByUpstreamUuid(String upstreamUuid) {
-            Criteria idCriteria = session.createCriteria(Owner.class, "Owner")
-                .setProjection(Projections.distinct(Projections.id()))
-                .createAlias("Owner.pools", "Pool")
-                .createAlias("Pool.product", "Prod", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("Prod.derivedProduct", "DProd", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("Prod.providedProducts", "PProd", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("DProd.providedProducts", "DPProd", JoinType.LEFT_OUTER_JOIN)
-                .add(Restrictions.or(
-                    CPRestrictions.in("Prod.id", productIds),
-                    CPRestrictions.in("DProd.id", productIds),
-                    CPRestrictions.in("PProd.id", productIds),
-                    CPRestrictions.in("DPProd.id", productIds)
-                ));
-
-            List<String> ownerIds = idCriteria.list();
-
-            if (ownerIds != null && !ownerIds.isEmpty()) {
-                DetachedCriteria criteria = DetachedCriteria.forClass(Owner.class)
-                    .add(CPRestrictions.in("id", ownerIds));
-
-                return this.cpQueryFactory.buildQuery(session, criteria);
-            }
+    public List<Owner> getOwnersWithProducts(Collection<String> productIds) {
+        if (productIds == null || productIds.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        CriteriaBuilder criteriaBuilder = this.entityManager.get().getCriteriaBuilder();
+        CriteriaQuery<Owner> query = criteriaBuilder.createQuery(Owner.class);
+        Root<Owner> owner = query.from(Owner.class);
+        query.distinct(true);
+        query.select(owner);
+
+        Join<Owner, Pool> pool = owner.join(Owner_.pools);
+        Join<Pool, Product> product = pool.join(Pool_.product, JoinType.LEFT);
+        Join<Product, Product> providedProducts = product.join(Product_.providedProducts, JoinType.LEFT);
+        Join<Product, Product> derivedProducts = product.join(Product_.derivedProduct, JoinType.LEFT);
+        Join<Product, Product> derivedProvidedProducts =
+            derivedProducts.join(Product_.providedProducts, JoinType.LEFT);
+
+        Expression<String> productExpr = product.get(Product_.id);
+        Expression<String> providedProductsExpr = providedProducts.get(Product_.id);
+        Expression<String> derivedProductsExpr = derivedProducts.get(Product_.id);
+        Expression<String> derivedPPExpr = derivedProvidedProducts.get(Product_.id);
+
+        // Block size reduced by 4 times to avoid exceeding the parameters limit as there
+        // are multiple IN clauses in the query below.
+        int blockSize = Math.min(this.getInBlockSize(), this.getQueryParameterLimit() / 4);
+        Set<Owner> owners = new HashSet<>();
+
+        for (List<String> block : Iterables.partition(productIds, blockSize)) {
+            List<Predicate> predicates = new LinkedList<>();
```

Comment:
Since we have a fixed number of predicates to add to the query, we can avoid the list->array juggling by just building it all at once:

```
Predicate predicate = criteriaBuilder.or(productExpr.in(block), providedProductsExpr.in(block), ..., ...);
query.where(predicate);
```

Reply:
Sure, done.